src/nupic/regions/SPRegion.cpp

```diff
@@ -0,0 +1,945 @@
+#include <nupic/ntypes/Value.hpp>
+#include <nupic/regions/SPRegion.hpp>
+#include <nupic/utils/Log.hpp>
+
+#define VERSION 1 // version for streaming serialization format
+
+namespace nupic {
+
+SPRegion::SPRegion(const ValueMap &values, Region *region)
+    : RegionImpl(region) {
+  // Note: the ValueMap gets destroyed on return so we need to get all of the
+  // parameters out of the map and set aside so we can pass them to the SpatialPooler
+  // algorithm when we create it during initialization().
+  args_.columnCount = values.getScalarT<UInt32>("columnCount", 0);
+  args_.potentialRadius = values.getScalarT<UInt32>("potentialRadius", 0);
+  args_.potentialPct = values.getScalarT<Real32>("potentialPct", 0.5);
+  args_.globalInhibition = values.getScalarT<bool>("globalInhibition", true);
+  args_.localAreaDensity = values.getScalarT<Real32>("localAreaDensity", -1.0f);
+  args_.numActiveColumnsPerInhArea = values.getScalarT<UInt32>("numActiveColumnsPerInhArea", 10);
+  args_.stimulusThreshold = values.getScalarT<UInt32>("stimulusThreshold", 0);
+  args_.synPermInactiveDec = values.getScalarT<Real32>("synPermInactiveDec", 0.008f);
+  args_.synPermActiveInc = values.getScalarT<Real32>("synPermActiveInc", 0.05f);
+  args_.synPermConnected = values.getScalarT<Real32>("synPermConnected", 0.1f);
+  args_.minPctOverlapDutyCycles = values.getScalarT<Real32>("minPctOverlapDutyCycles", 0.001f);
+  args_.dutyCyclePeriod = values.getScalarT<UInt32>("dutyCyclePeriod", 1000);
+  args_.boostStrength = values.getScalarT<Real32>("boostStrength", 0.0f);
+  args_.seed = values.getScalarT<Int32>("seed", 1);
+  args_.spVerbosity = values.getScalarT<UInt32>("spVerbosity", 0);
+  args_.wrapAround = values.getScalarT<bool>("wrapAround", true);
+  spatialImp_ = values.getString("spatialImp", "");
+
+  // variables used by this class and not passed on to the SpatialPooler class
+  args_.learningMode = (1 == values.getScalarT<UInt32>("learningMode", true));
+  args_.iter = 0;
+
+}
+
+SPRegion::SPRegion(BundleIO &bundle, Region *region) : RegionImpl(region) {
+
+  deserialize(bundle);
+}
+
+SPRegion::~SPRegion() {}
+
+void SPRegion::initialize() {
+  // Output buffers should already have been created diring initialize or deserialize.
+  Array &outputBuffer = getOutput("bottomUpOut")->getData();
+  NTA_CHECK(outputBuffer.getType() == NTA_BasicType_SDR);
+  UInt32 columnCount = (UInt32)outputBuffer.getCount();
+  if (columnCount == 0 || outputBuffer.getBuffer() == nullptr) {
+    NTA_THROW << "SPRegion::initialize - Output buffer not set.\n";
+  }
+
+  // All input links should have been initialized during Network.initialize().
+  // However, if nothing is connected it might not be. The SpatialPooler
+  // algorithm requires input.
+  //
+  // If there are more than one input link (FAN-IN), the input buffer will be the
+  // concatination of all incomming buffers.  
+  Input *in = getInput("bottomUpIn");
+  if (!in->hasIncomingLinks())
```

Comment:
can in be null?

Reply:
Only if the Spec is wrong, but yes we should check.  good catch.
I added a check.